iotlabcli/parser/main.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+
+# This file is a part of IoT-LAB cli-tools
+# Copyright (C) 2015 INRIA (Contact: admin@iot-lab.info)
+# Contributor(s) : see AUTHORS file
+#
+# This software is governed by the CeCILL license under French law
+# and abiding by the rules of distribution of free software.  You can  use,
+# modify and/ or redistribute the software under the terms of the CeCILL
+# license as circulated by CEA, CNRS and INRIA at the following URL
+# http://www.cecill.info.
+#
+# As a counterpart to the access to the source code and  rights to copy,
+# modify and redistribute granted by the license, users are provided only
+# with a limited warranty  and the software's author,  the holder of the
+# economic rights,  and the successive licensors  have only  limited
+# liability.
+#
+# The fact that you are presently reading this means that you have had
+# knowledge of the CeCILL license and that you accept its terms.
+
+"""Main parser."""
+
+# pylint:disable=too-few-public-methods
+
+import os
+import subprocess
+import sys
+from argparse import ArgumentParser, SUPPRESS, Action, REMAINDER
+
+from pkg_resources import iter_entry_points
+
+
+def main_print_help(parser):
+    """ Prints help for the main command """
+    parser.print_help()
+
+
+class CustomHelpAction(Action):
+    """ Ignoring --help if called after a subcommand is selected"""
+    def __init__(self, *args, **kwargs):
+        super(CustomHelpAction, self).__init__(*args, nargs=0, **kwargs)
+
+    def __call__(self, parser, args, values, option_string=None):
+        main_print_help(parser)
+
+        # no command or `help` command
+        if args.command == 'help':
+            parser.exit()
+
+
+def is_cmd(path, element):
+    """Whether a file is an iotlab subcommand."""
+    if element.startswith('iotlab-') \
+            and os.access(os.path.join(path, element), os.X_OK):
+        return True
+    return False
+
+
+def make_run_cmd(path, cmd):
+    """Closure to add the call to subprocess for each subcommand."""
+    absolute_path = os.path.join(path, cmd)
+
+    def run_cmd(args):
+        """Subcommand call with subprocess."""
+        subprocess.call([absolute_path] + args)
+
+    return run_cmd
+
+
+def get_iotlab_exec_subcommands():
+    """Get the dictionary of available iotlab subcommands in $PATH"""
+    commands = {}
+    for path in [p for p in sys.path
```

Comment:
This is overly complicated and suboptimal: you are looping other all executables available in the system path to find the iotlab commands. I would only use the know list of commands and if iotlab-ssh/serial_aggregator are installed (just use a try except), add the related commands to this list.

Reply:
Yeah, that would also work. But I think the posibility to have any subcommand be available without modifying cli-tools is powerful (just like git- subcommands, git also searches the whole system path).